src/codegen/dispatch.rs

```diff
@@ -358,3 +381,114 @@ pub(super) fn constructor_dispatch(
 }
+
+/// On Solana, prepare the data account after deploy; ensure the account is
+/// large enough and write magic to it to show the account has been deployed.
+fn solana_deploy(
+    contract_no: usize,
+    vartab: &mut Vartable,
+    cfg: &mut ControlFlowGraph,
+    ns: &mut Namespace,
+) {
+    // Make sure that the data account is large enough. Read the size of the
+    // account via `tx.accounts[0].data.length`.
+    let account_length = Expression::Builtin(
+        Loc::Codegen,
+        vec![Type::Uint(32)],
+        Builtin::ArrayLength,
+        vec![Expression::StructMember(
+            Loc::Codegen,
+            Type::DynamicBytes,
+            Expression::Subscript(
+                Loc::Codegen,
+                Type::Struct(StructType::AccountInfo),
+                Type::Array(
+                    Type::Struct(StructType::AccountInfo).into(),
+                    vec![ArrayLength::Dynamic],
+                ),
+                Expression::Builtin(
+                    Loc::Codegen,
+                    vec![Type::Array(
+                        Type::Struct(StructType::AccountInfo).into(),
+                        vec![ArrayLength::Dynamic],
+                    )],
+                    Builtin::Accounts,
+                    vec![],
+                )
+                .into(),
+                Expression::NumberLiteral(Loc::Codegen, Type::Uint(32), BigInt::zero()).into(),
+            )
+            .into(),
+            2,
+        )],
+    );
+
+    let is_enough = Expression::MoreEqual(
+        Loc::Codegen,
+        Box::new(account_length),
+        Box::new(Expression::NumberLiteral(
+            Loc::Codegen,
+            Type::Uint(32),
+            ns.contracts[contract_no].fixed_layout_size.clone(),
+        )),
+    );
+
+    let enough = cfg.new_basic_block("enough".into());
+    let not_enough = cfg.new_basic_block("not_enough".into());
+
+    cfg.add(
+        vartab,
+        Instr::BranchCond {
+            cond: is_enough,
+            true_block: enough,
+            false_block: not_enough,
+        },
+    );
+
+    cfg.set_basic_block(not_enough);
+
+    cfg.add(
+        vartab,
+        Instr::ReturnCode {
+            code: ReturnCode::AccountDataTooSmall,
+        },
+    );
+
+    cfg.set_basic_block(enough);
+
+    // Write contract magic number to offset 0
+    cfg.add(
+        vartab,
+        Instr::SetStorage {
+            ty: Type::Uint(32),
+            value: Expression::NumberLiteral(
+                pt::Loc::Codegen,
+                Type::Uint(64),
+                BigInt::from(ns.contracts[contract_no].selector()),
+            ),
+            storage: Expression::NumberLiteral(pt::Loc::Codegen, Type::Uint(64), BigInt::zero()),
+        },
+    );
+
+    // Calculate heap offset (align on 8 byte boundary)
+    let fixed_fields_size = ns.contracts[contract_no]
+        .fixed_layout_size
+        .to_u64()
+        .unwrap();
+
+    let heap_offset = (fixed_fields_size + 7) & !7;
```

Comment:
Just as a curiosity, what does this operation do?

Reply:
This aligns the heap on an 8 byte boundary. I'll add a comment